pkg/controller/attach_detach.go

```diff
@@ -312,64 +311,69 @@ func (r *ReconcileAttachDetach) triggerDetach(ctx context.Context, azVolumeAttac
+		// Update state to detaching
+		_, derr := updateState(azv, azdiskv1beta2.Detaching, normalUpdate)
+		return derr
+	}
 
-		var updatedObj client.Object
-		if updatedObj, err = azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRIStatus); err != nil {
-			return err
-		}
-		azVolumeAttachment = updatedObj.(*azdiskv1beta2.AzVolumeAttachment)
-
-		w.Logger().V(5).Info("Detaching volume")
-		waitCh := make(chan goSignal)
-		//nolint:contextcheck // call is asynchronous; context is not inherited by design
-		go func() {
-			var detachErr error
-			_, goWorkflow := workflow.New(ctx)
-			defer func() { goWorkflow.Finish(detachErr) }()
-			waitCh <- goSignal{}
-
-			goCtx := goWorkflow.SaveToContext(context.Background())
-			cloudCtx, cloudCancel := context.WithTimeout(goCtx, cloudTimeout)
-			defer cloudCancel()
-
-			var updateFunc func(obj client.Object) error
-			updateMode := azureutils.UpdateCRIStatus
-			detachErr = r.detachVolume(cloudCtx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName)
-			if detachErr != nil {
-				updateFunc = func(obj client.Object) error {
-					azv := obj.(*azdiskv1beta2.AzVolumeAttachment)
-					azv = updateError(azv, detachErr)
-					_, derr := updateState(azv, azdiskv1beta2.DetachmentFailed, forceUpdate)
-					return derr
-				}
-				//nolint:contextcheck // final status update of the CRI must occur even when the current context's deadline passes.
-				if _, uerr := azureutils.UpdateCRIWithRetry(goCtx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.ForcedUpdateMaxNetRetry, updateMode); uerr != nil {
-					w.Logger().Errorf(uerr, "failed to update final status of AzVolumeAttachement")
-				}
-			} else {
-				//nolint:contextcheck // delete of the CRI must occur even when the current context's deadline passes.
-				if derr := r.cachedClient.Delete(goCtx, azVolumeAttachment); derr != nil {
-					w.Logger().Error(derr, "failed to delete AzVolumeAttachment")
-				}
-			}
-		}()
-		<-waitCh
+	var updatedObj client.Object
+	if updatedObj, err = azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRIStatus); err != nil {
+		return err
 	}
+	azVolumeAttachment = updatedObj.(*azdiskv1beta2.AzVolumeAttachment)
+
+	w.Logger().V(5).Info("Detaching volume")
+	waitCh := make(chan goSignal)
+	//nolint:contextcheck // call is asynchronous; context is not inherited by design
+	go func() {
+		var detachErr error
+		_, goWorkflow := workflow.New(ctx)
+		defer func() { goWorkflow.Finish(detachErr) }()
+		waitCh <- goSignal{}
+
+		goCtx := goWorkflow.SaveToContext(context.Background())
+		cloudCtx, cloudCancel := context.WithTimeout(goCtx, cloudTimeout)
+		defer cloudCancel()
+
+		var updateFunc func(obj client.Object) error
+		updateMode := azureutils.UpdateCRIStatus
```

Comment:
I think we can get rid of the `updateMode` variable here and just feed `UpdateCRIStatus` directly to `UpdateCRIWithRetry` function.